Develop

llanaba opened the pull request.

llanaba committed:
add class FormValidator

NotACat1 committed:
fix: correction of the Section class

1. renderItems - fixing the method so that it creates and returns a ready-made markup.

llanaba committed:
create class Card

llanaba committed:
link class Card to index.js

llanaba committed:
update class User, link class User to index.js

llanaba committed:
modify index.js

llanaba committed:
update Card class, add: rendering likes and bin

llanaba committed:
add functionality: like/dislike cards

llanaba committed:
Merge pull request #20 from llanaba/card

Card

llanaba committed:
add delete card functionality

NotACat1 committed:
Update README.md

llanaba committed:
Merge pull request #21 from llanaba/README

Update README.md

llanaba committed:
add: create new card

llanaba committed:
reverts "add delete card functionality"

This reverts commit 1777d38175c5c664682178f9d88bdbb5a4ac1ffb.

llanaba committed:
Merge branch 'card' of github.com:llanaba/mesto-project into card

llanaba committed:
merging popup and card

llanaba committed:
put together popups and index

llanaba committed:
add profile update functionality

NotACat1 committed:
docs: adding comments

NotACat1 committed:
build: adding card deletion functionality

- сomments have been added
- errors and bugs have been fixed
- the functionality for deleting a photo card has also been added

NotACat1 committed:
build: adding card deletion functionality

- сomments have been added
- errors and bugs have been fixed
- the functionality for deleting a photo card has also been added

NotACat1 committed:
docs: adding comments

NotACat1 committed:
Merge branch 'card' of https://github.com/llanaba/mesto-project into card

NotACat1 committed:
docs: adding comments

NotACat1 committed:
docs: adding comments

NotACat1 committed:
refactor: removed unnecessary message outputs to the console

NotACat1 committed:
perf: убраны лишние действия

NotACat1 committed:
build: adding a class for a confirmation form

NotACat1 committed:
ci: correction of errors in the work

- the card is displayed after being added
- confirmation of photo card deletion

NotACat1 committed:
docs: adding comments

llanaba committed:
fix: order of rendering cards

NotACat1 committed:
Merge pull request #22 from llanaba/card

Card

gennady-bars commented on src/components/api.js:
```diff
@@ -26,39 +28,35 @@ export class Api {
       headers: this._headers
     })
       .then((res) => {
         return this._getResponseData(res);
```
**Можно лучше** 

Можно сделать **универсальный метод запроса с проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе:

```js

  _request(url, options) {
    return fetch(url, options).then(this._getResponseData)
  }

```

И теперь просто нужно заменить все `fetch` на `this._request`, убрав дублирование проверки на `ok`. Все остальное будет без изменений 

И даже можно поместить внутрь `this._baseUrl`, чтобы не дублировать его в каждом запросе. Тогда нужно будет передавать просто `endpoint` в вызов

gennady-bars commented on src/components/Card.js:
```diff
@@ -0,0 +1,102 @@
+    // we draw something that does not depend on the user
+    this._element.cardName = this._element.querySelector(cardSelectors.cardName);
+    this._element.cardImage = this._element.querySelector(cardSelectors.cardImage);
+    this._element.likeCounter = this._element.querySelector(cardSelectors.likeCounter);
```
Отлично, что нашли элементы `DOM` 1 раз тут и сделали их полями класса.

gennady-bars commented on src/components/FormValidator.js:
```diff
@@ -0,0 +1,69 @@
+      return !inputElement.validity.valid;
+    });
+  }
+
```
**Можно лучше**

Можно сделать отдельный  метод для очистки ошибок и управления кнопкой `resetValidation`.

В нем Вы будете пробегаться по ` this._inputList` и очищать ошибки с инпутов.

Это нужно потому, что ошибки остаются в инпутах, если пользователь начал что-то вводить в инпут неправильное, потом передумал и закрыл попап. А при повторном открытии **ошибки** в инпутах **остались**, так как они не очищаются при открытии сейчас. Вот нужно сделать это очищение и управление кнопкой (она должна быть неактивной в попапе `добавления новой карточки`).

Передавать в этот метод ничего не нужно, так как Вам нужно сделать `отдельные экземпляры` класса `FormValidator` для разных попапов с формой, и эти экземпляры будут знать свои инпуты и кнопку.

Вызывать resetValidation нужно в том месте, где Вы обрабатываете нажатие на **кнопку открытия** попапа с формой.

Это будет выглядеть примерно так:

```js
 resetValidation() {
      this._toggleButtonState(); <== управляем кнопкой ==

      this._inputList.forEach((inputElement) => {
        this._hideError(inputElement) <==очищаем ошибки ==
      });

    }

```

gennady-bars commented on src/components/Popup.js:
```diff
@@ -19,24 +20,24 @@ export default class Popup {
   // adding event handlers
   setEventListeners () {
-    this._selector.addEventListener('keydown', (evt) => {
+    document.addEventListener('keydown', (evt) => {
       this._handleEscClose(evt);
     });
```
Это нужно удалить из `setEventListeners`


По заданию и чек-листу обработчик Escape должен правильно удаляться при закрытии попапа

В таком виде невозможно удалить никакой обработчик, так как метод `bind` всегда создает `новую` ссылку на функцию, в итоге движок браузера просто не сможет найти функцию, чтобы ее удалить. Ошибки никакой браузер не выдаст, так как ему без разницы, что он не нашел функцию.

Для того чтобы ссылка стала постоянной, нужно "привязать" функцию к контексту `this` в конструкторе 1 раз с помощью `bind`.

```js
 constructor(popupSelector) {
    this._popup = …………….
    this._handleEscClose = this._handleEscClose.bind(this) <==== привязали 1 раз ===
  }


```

и теперь при открытии навешиваете и передаете только `ссылку`:


```js
 open() {
    открыли попап
    document.addEventListener('keydown', this._handleEscClose); <=== только cсылку передаете ==
    
  }

```

и удаляете при закрытии точно с такой же ссылкой:

```js
 close() {
    закрыли попап
    document.removeEventListener('keydown', this._handleEscClose); <==== такая же ссылка должна быть
  }


```

Подробнее про правильное удаление обработчиков тут: https://learn.javascript.ru/introduction-browser-events#addeventlistener

gennady-bars commented on src/components/PopupWithConfirmation.js:
```diff
@@ -0,0 +1,30 @@
+  }
+  
+  // setting an item to delete
+  callBackDeleteItem (data) {
```
**Названия** методов и функций нужно начинать **с глагола** в начальной форме (такой пункт есть в чек-листе). Это обычная международная практика, чтобы функция говорила, что она делает. **Все функции в проекте нужно исправить, если не начинаются с глагола** 

 `setSubmitHandler`

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -1,29 +1,38 @@
-    this._inputList = this._selector.querySelectorAll('.form__input');
-    
+  _getInputValues () {
+    this._inputList = [...this._form.querySelectorAll('.form__input-text')];
```
Все элементы, которые используются в нескольких разных методах классов или поиск которых осуществляется при каждом срабатывании метода, желательно найти **1 раз** и объявить их полями класса (**классовыми** переменными с `this`). Чаще всего это удобно сделать в `конструкторе`, `generateCard` или `_setEventListeners`, и тогда эти переменные будут **доступны** по всему коду класса. Это позволит сэкономить время на повторный поиск элементов, так как они 1 раз были найдены. И не надо будет их больше передавать в вызовы методов. И в **портфолио** такой код будет выгодно смотреться для работодателя.

Эти элементы обычно `this._inputList`, `this._submitButton`, `this._form`, `this._cardImage`, `this._likeButton`

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -1,29 +1,38 @@
+      if (rezSubmit) {
+        rezSubmit
+          .then((rez) => {
+            this._form.reset();
```
`            this._form.reset();
`

Это тут не нужно, так как уже есть внутри ` this.close();`

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -1,29 +1,38 @@
-    
     return this._formValues;
   }
 
```
**Можно лучше**

Лучше сделать метод `renderLoading` в классе `PopupWithForm`. Там найти 1 раз кнопку сабмита в конструкторе и менять ей текст.

```js
  constructor({ popupSelector, handleSubmit }) {
    super(popupSelector);
    // фиксируем начальный текст кнопки 1 раз в конструкторе
    this._submitBtnText = this._submitBtn.textContent
  }
  // указываем 2 параметра (2й с текстом по умолчанию, чтобы не указывать лишний раз его)
  renderLoading(isLoading, loadingText='Сохранение...') {
    if (isLoading) {
      this._submitBtn.textContent = loadingText;
    } else {
      this._submitBtn.textContent = this._submitBtnText;
    }
  }

```

И теперь не нужно даже думать, как возвращать начальный текст кнопке. Все будет работать для любого текста

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -1,29 +1,38 @@
+      }
     });
   }
 
```
**Можно лучше**

Можно сделать метод ` setInputValues` в классе `PopupWithForm`, который будет вставлять данные в инпуты:

```js

  setInputValues(data) {
    this._inputList.forEach((input) => {
      // тут вставляем в `value` инпута данные из объекта по атрибуту `name` этого инпута
      input.value = data[input.name];
    });
  }


```

И не нужно будет искать эти инпуты в `index.js` и что-то вставлять в них при открытии профиля.

gennady-bars commented on src/components/PopupWithImage.js:
```diff
@@ -1,20 +1,18 @@
-      super.open();
-    }
+
+    this._image.onload = super.open();
```
`    this._image.onload = super.open();
`

Это так не работает


Нужно просто 


```js

    super.open();

```

gennady-bars commented on src/components/Section.js:
```diff
@@ -1,20 +1,23 @@
-  // }
+  // clearing the container of all photo cards
+  clear() {
+    [...this._container.querySelectorAll('.card')].forEach((card) => card.remove());
```
Не нужно искать все карточки. 


Очищается секция очень легко


```js
  clear() {
    this._container.innerHtml = ""
  }

```


В таких случаях можно использовать `innerHtml`, так как обнуляется все содержимое

gennady-bars commented on src/components/UserInfo.js:
```diff
@@ -1,33 +1,66 @@
+  }
+
+  // fetches user's info from server, processes, renders and returns it
+  getUserInfo () {
+    const info = this._getInfoApi();
+    info.then((userInfo) => {
+      this._processUserInfo(userInfo);
+      this._renderUserInfo();
+    })
+    .catch((err) => {
+      console.log(err);
+    });
+    return info;
+  }
+
+  // unpacks user's info received from server and sets user's attributes
+  _processUserInfo ({ name, about, avatar, _id }) {
```
Разработчики задания ошиблись. Нельзя использовать `api` внутри класса `UserInfo`, иначе невозможно сделать правильный и аккуратный код (без дублирований). Это должны скоро исправить, но… все должно было описано вот так:

```js

class UserInfo {
  // теперь все данные пользователя устанавливаются тут, включая аватар
  constructor({ nameSelector, jobSelector, avatarSelector }) {
    this._nameElement = document.querySelector(nameSelector);
    this.jobElement = document.querySelector(jobSelector);
    this._avatarElement = document.querySelector(avatarSelector);
  }

  // метод getUserInfo должен возвращать данные из профиля, взяв их из `textContent`. 
  getUserInfo = () => {
    return {
      name: this._nameElement.textContent,
      about: this.jobElement.textContent,
      userId: this._userId,  // _id тоже можно тут получать, чтобы испольозовать в `index.js` для создания карточек
    };
  };
  // метод `setUserInfo` должен получать в вызов все данные пользователя и устанавливать их внутри
  setUserInfo({ name, about, avatar, _id }) {
    this._nameElement.textContent = name;
    this.jobElement.textContent = about;
    this._avatarElement.src = avatar;
    this._userId = _id;
  }
}

```

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,171 +1,146 @@
+const editProfilePopup = new PopupWithForm ( 
+  popupSelectors.editProfile,
+  validate,
+  { submit: user.setUserInfo.bind(user) }
```
**Можно лучше**


Если интересно, посмотрите, как можно избавиться от дублирования изменения текста кнопки сабмита, отлова ошибок и закрытия попапа в каждом запросе:

```js

// можно сделать универсальную функцию, которая принимает функцию запроса,  экземпляр попапа и текст во время загрузки (опционально)
function handleSubmit(request, popupInstance, loadingText = "Сохранение...") {

  // изменяем текст кнопки до вызова запроса
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // закрывать попап нужно только в `then`
      popupInstance.close()
    })
    .catch((err) => {
      // в каждом запросе нужно ловить ошибку
      console.error(`Ошибка: ${err}`);
    })
    // в каждом запросе в `finally` нужно возвращать обратно начальный текст кнопки
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}

// пример оптимизации обработчика сабмита формы профиля
function handleProfileFormSubmit(inputValues) {
  // создаем функцию, которая возвращает промис, так как любой запрос возвращает его 
  function makeRequest() {
    // `return` позволяет потом дальше продолжать цепочку `then, catch, finally`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // вызываем универсальную функцию, передавая в нее запрос, экземпляр попапа и текст изменения кнопки (если нужен другой, а не `"Сохранение..."`)
  handleSubmit(makeRequest, profilePopup);
}

```

В итоге, у Вас исчезнет огромное кол-во дублирования логики из кода.

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,171 +1,146 @@
+
+// creating a photo card and placing it in storage
+function renderPhotoCard (photoCardItem) {
+  const card = new Card (
```
Для создания новой карточки нужно сделать отдельную функцию `createCard` , чтобы не дублировать код и разделить логику вставки: она будет возвращать готовую карточку с уже установленными обработчиками через `return`, а вставлять в DOM там **не нужно**.
Это нужно для того, чтобы можно было разделить логику вставки и спрятать логику создания (в 5-6 спринтах карточка создается с помощью клонирования, потом (7-9) с помощью класса `Card`, а потом еще как-то в будущем). Можно же вставлять карточку в начало с `prepend`, в конец с `append`, вообще не вставлять, а сделать массив готовых карточек, а потом уже вставить всё разом в DOM. И такой пункт есть в чек-листе. Скрин из него http://joxi.ru/a2XNb9QtQj3Eer 

На всякий случай повторюсь: в функции `createCard` вставлять карточку в DOM  **не нужно**.

```js		

function createCard(item) {
    // тут создаете карточку и возвращаете ее
  return cardElement.generate(user)
}

```

тут останется 2 строчки кода:

```js

  const cardElement = createCard(item)

  initialCardList.addItem(cardElement);
```

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,171 +1,146 @@
-      let [userData, cardsData] = values;
-      renderProfileOnPage(userData);
-      renderInitialCards(userData._id, cardsData);
+    .then(([ userData, cardsData ]) => {
```
Как раз `userData` нужно использовать для установки данных пользователя. Это не должно быть где-то еще

Только 1 раз делается запрос данных пользователя на сервер

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,171 +1,146 @@
+// Enabling validation for all forms on the site
+// relation to the form validation class (FormValidator)
+function validate(form) {
+  const validator = new FormValidator(validationSelectors, form);
```
**Можно лучше**

Если будет интересно, можно универсально создать экземпляры валидаторов всех форм, поместив их все в один объект, а потом брать из него валидатор по атрибуту `name`, который задан для формы. Это очень универсально и для любого кол-ва форм подходит.

```js
const formValidators = {}

// Включение валидации
const enableValidation = (config) => {
  const formList = Array.from(document.querySelectorAll(config.formSelector))
  formList.forEach((formElement) => {
    const validator = new FormValidator(config, formElement)
// получаем данные из атрибута `name` у формы
    const formName = formElement.getAttribute('name')

   // вот тут в объект записываем под именем формы
    formValidators[formName] = validator;
    validator.enableValidation();
  });
};

enableValidation(config);

```

И теперь можно использовать валидаторы для деактивации кнопки и тд

```js

formValidators[ profileForm.getAttribute('name') ].resetValidation()

// или можно использовать строку (ведь Вы знаете, какой атрибут `name` у каждой формы)
formValidators['profile-form'].resetValidation()

```

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,171 +1,146 @@
+// Creating New Card popup
+createCardPopup.setEventListeners(); // enabling event handlers
+buttons.addCard.addEventListener('click', (evt) => {
+  createCardPopup.open(); // opening of the popup
```
* В попапе добавления новой карточки нужно **деактивировать** кнопку сабмита, если инпуты пустые, чтобы не было возможности нажать на кнопку и добавить пустую карточку без картинки и названия и чтобы показать пользователю, что данные пока невалидны. А сейчас после 1го добавления карточки кнопка активируется и её нужно деактивировать опять. Кнопка должна быть **серой** (как при валидации) и с атрибутом `disabled`.
* Теперь **только** методы класса `FormValidator` должны активировать / деактивировать кнопку сабмита и очищать ошибки в `index.js`

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,171 +1,146 @@
-  handleAddCardClick(popupAddCard);
+changeAvatarPopup.setEventListeners(); // enabling event handlers
+buttons.changeAvatar.addEventListener('click', (evt) => {
+  changeAvatarPopup.open(); // opening of the popup
```
* В попапе добавления новой карточки нужно **деактивировать** кнопку сабмита, если инпуты пустые, чтобы не было возможности нажать на кнопку и добавить пустую карточку без картинки и названия и чтобы показать пользователю, что данные пока невалидны. А сейчас после 1го добавления карточки кнопка активируется и её нужно деактивировать опять. Кнопка должна быть **серой** (как при валидации) и с атрибутом `disabled`.
* Теперь **только** методы класса `FormValidator` должны активировать / деактивировать кнопку сабмита и очищать ошибки в `index.js`

И в аватаре тоже

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная, но есть некоторые недочеты:




* Все файлы с классами по чек-листу нужно сделать с **большой** буквы. B `api.js` в том числе
* Все черные сердечки опять становятся белыми после перезагрузки страницы
* В попапе добавления новой карточки нужно **деактивировать** кнопку сабмита, если инпуты пустые, чтобы не было возможности нажать на кнопку и добавить пустую карточку без картинки и названия и чтобы показать пользователю, что данные пока невалидны. А сейчас после 1го добавления карточки кнопка активируется и её нужно деактивировать опять. Кнопка должна быть **серой** (как при валидации) и с атрибутом `disabled`.
* Теперь **только** методы класса `FormValidator` должны активировать / деактивировать кнопку сабмита и очищать ошибки в `index.js`
* Нужно **изменять текст** кнопки сабмита при ожидании ответа от сервера. И это должно быть видно пользователю.
* По заданию и чек-листу обработчик Escape должен правильно удаляться при закрытии попапа
* **Названия** методов и функций нужно начинать **с глагола** в начальной форме (такой пункт есть в чек-листе). Это обычная международная практика, чтобы функция говорила, что она делает. **Все функции в проекте нужно исправить, если не начинаются с глагола** 
* Все элементы, которые используются в нескольких разных методах классов или поиск которых осуществляется при каждом срабатывании метода, желательно найти **1 раз** и объявить их полями класса
* Для создания новой карточки нужно сделать отдельную функцию `createCard` , чтобы не дублировать код и разделить логику вставки: она будет возвращать готовую карточку с уже установленными обработчиками через `return`, а вставлять в DOM там **не нужно**.
* При открытии попапа профиля нужно **вставлять** в его инпуты данные пользователя, которые отображены сейчас на сайте, чтобы можно было их отредактировать сразу и обновить. 



Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,171 +1,146 @@
+// Profile Editing popup (info and avatar)
+editProfilePopup.setEventListeners(); // enabling event handlers
+buttons.editProfile.addEventListener('click', (evt) => {
+  editProfilePopup.open(); // opening of the popup
```
При открытии попапа профиля нужно **вставлять** в его инпуты данные пользователя, которые отображены сейчас на сайте, чтобы можно было их отредактировать сразу и обновить. 


Для этого есть в классе `UserInfo` метод `getUserInfo`

llanaba committed:
refactor Api class

llanaba committed:
refactor validation

llanaba committed:
fix setting Event Listener for Esc close popup

llanaba committed:
refactor popups

llanaba committed:
fix: UserInfo class

NotACat1 committed:
style: cosmetic changes

- added missing characters ';'
- added a comment to the '_request' method

NotACat1 committed:
perf: improved operation of the method '_userLikesCard'

improved the operation of the method of checking the card's like by the user

NotACat1 committed:
perf: fields of the submit button have been removed

removed the search for the submit button element. Since this button is only needed in the PopupWithForm class.

NotACat1 committed:
build: adding fields for the submit button

Adding fields for the submit button. Also added comments to the methods of the class.

NotACat1 committed:
fix: bug with the display of the previous image

Fixing a bug related to the display of the previous open image.
Solution: when closing the popup, delete the fields with information about the previous image (placed in the '_resetInfo' method).

There is also a solution via '.onload' images. Solution: open the popup only when the image is loaded. Lines - 23-27

NotACat1 committed:
style: the ';' character is missing

NotACat1 committed:
style: improved code readability

Improved code readability in the class constructor. Also removed the extra "console.log"

NotACat1 committed:
build: added constants: error time and form names

1. timeErr - time in seconds, how long the submit error is displayed;
2. nameForms - object with form names

NotACat1 committed:
ci: bug fixes for the script

- working out the logic of the buttons when the 'submit' event occurs (lines: 30, 33, 49, 52, 67, 69-70)
- adding and using new constants (lines: 5-6)
- dependency fixes (lines: 115-118)
- cosmetic changes (lines: 48, 85-87, 134, 139-140, 158-160)

NotACat1 committed:
perf: error output to the console

NotACat1 committed:
perf: error output to the console

NotACat1 committed:
refactor: processing a request to the server is output to index.js

NotACat1 committed:
refactor: changing the functionality this._renderer

Creating an element + placing it in a container -> creating an element

NotACat1 committed:
perf: rethinking and generalizing the work of submit

- Output of the handleSubmit functionality to separate functions.
- renderPhotoCard -> createCard. The functionality of adding cards to the storage has been removed, only the creation of a card element remains.
- Added comments to the code.
- Constants and functions are lifted up by the file.

NotACat1 committed:
Update README.md

llanaba committed:
brush up

llanaba committed:
Merge pull request #24 from llanaba/fix/review1

Fix/review1

llanaba committed:
Rename api.js to Api.js

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.

NotACat1 committed:
Update README.md
